### DISCUSSION TITLE
[A-L] (23 ottobre 2023) Esercizio 1

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#79] [Esami A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/d42729ee-8a42-4d5d-accf-905b9da11018)
(Classica distanza tra sottografi)

**Comment by Frisbon** [comment]
```py
def BFS_sottog(G,X,Y):
    D = [-1,...,-1]; #-1 implica non visitato/calcolato
    C = Queue();
    for x in X: 
        D[x] = 0
        C.enq(x)
    
    while C != empty:
        y = C.deq()
        if y.out != empty:
            z = y.out[0]
            if D[z] < 0: #se non visitato
                D[z] = D[y] + 1
                C.enq(z)
                
    m = -999
    for y in Y: m = min(m, D[y])
    return m
```

**Reply by Wichele13** [reply]
ma S cos'è?

**Reply by Frisbon** [reply]
Perdonami, inizialmente avevo usato lo Stack, solo dopo ho deciso di usare un BFS e implementare una coda al suo posto. Ho corretto il codice

**Comment by cornflexxx** [comment]
Prendo tutti gli archi che vanno da un nodo di $X$ ad un altro nodo di $X$ e modifico il peso a $w(u,v)= 0$, successivamente modifico il peso di  tutti gli archi che precedentemente non ho considerato con $w(p,q) = 1$  con $(p,q)\in E(G)- E(X)$


```
BFS_modificata(X,Y,G):

	D = [-1,..,-1]
	Q : Coda
	for (v,u) in E(X) :
		w(v,u) = 0
	for (v,u) in E(G)-E(X) :
		w(v,u) = 1
	v :=  v in X
	Q.add(v)
	while !isEmpty(Q):
		x = Q.pop()
		for u in N(x):
			if D[u] == -1:
				D[u] = D[x] + dist(x,u)
				Q.ass(u)
	min = +infinito
	for u in Y:
		if  d[v] < min:
			min = d[v]
	return min
```

**Reply by 18xChris** [reply]
Non capisco una cosa:
Avendo il seguente grafo
![image](https://github.com/user-attachments/assets/9b54f49d-fb61-42f7-9bbc-d8a14e73e874)
Il tuo algoritmo non dovrebbe comportarsi nel seguente modo?
Nel caso parta dall'ultimo vertice di sx (in basso)  lui arriva nel primo nodo not in X U Y e da li' parte tutta una bfs
Ma questa bfs va a modificare i costi su tutti i nodi not in X U Y, per cui quando arrivera' al nodo che sta alla dx del nodo di X piu' in alto, il suo costo non sara' 0 (dovrebbe essere 0 perche' e' adiacente ad un nodo di X) 
E quindi il costo finale sara' errato
Sbaglio?

**Reply by 18xChris** [reply]
Come farei io:

```Python
def bfsm(G,x,y,X,Y):
  costi = [0..0]
  vis = [0..0]
  gruppo = [-inf..-inf]

  for v in X:
    gruppo[v] = 1

  for v in Y:
    if gruppo[v] == 1:
      return 0
    gruppo[v] = 0
    


  newNode = Node() 
  for v in X:
    for vv vicino di v:
      if gruppo[vv] == -inf:
          newNode.vicini.append(vv)
      if gruppo[vv] = 0:
          return 1
  
  Q eueue
  Q.push(newNode)
  c = 0
  distanze = [0..0]
  while Q != None:
    v = Q.pop()
    for vv in v.vicini:
      if vis[vv] == 0:
        if gruppo[vv] == 0:
          return distanze[v]+1
        distanze[vv]=distanze[v]+1
        vis[vv] = 1
        Q.push(vv)


  return -1 #|Y| = 0
```

**Comment by IonutCicio** [comment]
L'idea è visitare il grafo con un normale BFS partendo dai nodi di $X$. Se il nodo successivo nella visita appartiene a $X$, la distanza è impostata a 0, altrimenti aumenta di 1. Se il nodo di cui si è appena calcolata la distanza appartiene a $Y$, ritorno la distanza calcolata.

```rust
fn subgraphs_distance(graph: &[Vec<usize>], origin: &[bool], destination: &[bool]) -> Option<usize> {
    let mut queue: VecDeque<_> = origin.iter().enumerate().filter(|(_, &x)| x).map(|(i, _)| i).collect();
    let mut distance: Vec<Option<usize>> = vec![None; graph.len()];

    while let Some(x) = queue.pop_back() {
        for &v in &graph[x] {
            if distance[v].is_none() {
                queue.push_back(v);
                distance[v] = Some(distance[x].unwrap_or(0) + if origin[v] { 0 } else { 1 });

                if destination[v] {
                    return distance[v];
                }
            }
        }
    }

    None
}
```

**Comment by CiottoloMaggico** [comment]
![immagine](https://github.com/user-attachments/assets/9b291bfd-c542-433d-bd3a-0be197c93905)

**Comment by cek753** [comment]
<img width="935" height="1413" alt="image" src="https://github.com/user-attachments/assets/2d5eaf2a-8a24-4127-a103-abe34a737bf5" />
